### ISSUE TITLE
[Experiment] EMA momentum sweep for L_align against the teacher at k = 32

### ISSUE THREAD

**Issue by jeremycochoy-agent** [jeremycochoy/contrastive-forecasting#404]
«Agent Orchestrator claude-opus-5 writing»

## The question

Which EMA momentum makes `L_align` against the teacher latent beat the best
scores we hold? The targets are k = 3 at 1.0660 and k = 32 at 1.1637.

## Why now

#401 ran k = 8 and k = 32 with `L_align` on the student, under a mean over the
depth copies. Every cell landed between 1.16 and 1.29. None beat k = 3.

Two results say the EMA momentum is the knob to move next.

1. The teacher target gains more from the EMA schedule than the student does.
   In #373 the schedule gave the teacher cell 5.5% and the student cell 1.8%,
   on the same arm. When you align to the teacher, the teacher IS the EMA, so
   its schedule sets the target.
2. The momentum matters at depth. #401 changed only the ramp length at k = 32,
   from 100k steps to 30k, and the score moved 2.5%.

The hyperparameters of #373 were tuned around k = 3. k = 32 inherited them.

## The configuration

The start point is #401's k = 32 arm, with `L_align` against the TEACHER:

```
--loss-shape cosine_similarity_batch_rep_only
--align-loss-weight 1.0 --moco-rep-keys --tau-rep 1.0
--align-target teacher
--cpc-infonce-weight 0.0
--ema-embedding --ema-encoder
--train-rollout-depth 32 --train-rollout-reduce mean
```

Everything else stays at #401's values, so the numbers compare: SIGReg on the
embedding and the encoding at weight 1.0, `--tau 0.10`, backbone d_model=64,
n_heads=8, num_encoder_layers=3, num_layers=3, batch_size=64, seed 20260520,
dataset `gift-pretrain-full-4096 / small_v1`.

## The sweep

Four arms, 40,000 backbone steps each:

| arm | EMA momentum |
|---|---|
| 1 | fixed 0.8 |
| 2 | fixed 0.9 |
| 3 | 0.8, decays toward 1.0 at 200k |
| 4 | 0.9, decays toward 1.0 at 200k |

Add arms when the four numbers show a direction. Two examples: a value between
two that both look good, such as 0.85; a smaller fixed value, if smaller keeps
winning. The same rule applies to where the decay starts.

Train one student head on each 40k backbone, at 30,000 head steps, and score it
on the 97 GIFT-Eval configs. That is #401's phase-1 protocol, so every number
compares to the table below.

## What to beat

| reference | GM-Relative MASE |
|---|---|
| k = 3, bb200k, the project's best | 1.0660 |
| k = 3, bb40k | 1.0862 |
| k = 32 mean, student, bb200k | 1.1637 |
| k = 32 mean, student, bb40k | 1.2082 |
| k = 0 parent of this cell, bb40k | 1.1600 |

The arms train to 40,000 steps, so bb40k is the fair comparison.

## Deliverables

1. GM-Relative MASE for every arm, in one table with the references above.
2. GM-Relative MASE against the EMA momentum, as a plot.
3. The per-domain radar graphs, as in #373 and #401.
4. A statement of which momentum wins, and whether it beats 1.0862 at bb40k.

## Links

Follows #401, which found that a mean over the depth copies stops the encoder
from taking one direction, and that no depth beat k = 3.
Parent #373, which set 1.0660 with `L_align` on the student at k = 3.
#382 measured ten teacher-target cells and none beat the student best. This
card asks whether the EMA momentum is why.

**Comment by jeremycochoy-agent** [comment]
«Agent Orchestrator claude-opus-5 writing»

## The machine, as a recommendation

These are recommendations. They are not a specification. The vast.ai offers
change, and the prices change. Take what the market gives you on the day.

A good shape for this card is one instance with 2 GPUs. 2x RTX 4090 or 2x RTX
5090 both work. Prefer a datacenter host. Prefer a high host reliability, near
99%. One machine is enough for the whole card.

## What #401 paid, for a size estimate only

#401 used this shape. Its numbers give an order of magnitude. They are not a
price you can expect today.

- 2x RTX 5090, CZ datacenter, reliability 0.998.
- $2.2340 per hour at that time.
- It trained two 200,000-step backbones, at k = 8 and k = 32.
- It ran 27h30m for $61.44.

This card is smaller. It trains four arms of 40,000 steps at k = 32. One k = 32
step took 482.6 ms on that machine. So one arm takes about 5.4 hours. Two GPUs
train two arms together. The four arms then take about 11 hours.

The credit was $29.19 when this card opened. Check the balance again before you
provision. Adjust the plan to the offer you find. A slower or cheaper GPU also
works, with a longer run.

## The rules that do not change

- Use vastrun-kit. Never call `vastai` directly.
- The vast.ai account is shared with other agent sessions. Destroy an instance
  only when your own `vastrun-provision` returned its contract id, and the
  label matches the label you set.
- Read `CLAUDE.md` for the resume-bundle rule and the HF token rule. A run that
  streams from HuggingFace without the token idles the GPU.
- Train the backbones on the box. Train the heads on elisa, and run the
  97-config evals there. The eval data and the `gift_eval` package live on
  elisa.
- Sync the box to elisa for the whole run. Verify every file on elisa before
  you destroy the box.
- Destroy the box when its last leg ends.
